tests/tracer/test_processors.py

```diff
@@ -216,3 +217,84 @@ def test_aggregator_partial_flush_2_spans():
     assert writer.pop() == [child1, child2]
     parent.finish()
     assert writer.pop() == [parent]
+
+
+def test_trace_top_level_span_processor_same_service_name():
+    """Parent span and child span have the same service name"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "parent", service="service")
+    parent._local_root = parent
+
+    child = Span(None, "child", service="service")
+    child.parent_id = parent.span_id
+    child._parent = parent
+    child._local_root = parent
+
+    trace_processors.process_trace([parent, child])
+    # parent span should be set to top_level
+    # child span should not contain top_level metric
+    assert parent.get_metric("_dd.top_level") == 1
+    assert "_dd.top_level" not in child.metrics
+
+
+def test_trace_top_level_span_processor_different_service_name():
+    """Parent span and child span have the different service names"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "parent", service="service")
+    parent._local_root = parent
+
+    child = Span(None, "child", service="new_service_name")
+    child.parent_id = parent.span_id
+    child._parent = parent
+    child._local_root = parent
+
+    trace_processors.process_trace([parent, child])
+    # child span should be set to top_level
+    assert parent.get_metric("_dd.top_level") == 1
+    assert child.get_metric("_dd.top_level") == 1
+
+
+def test_trace_top_level_span_processor_orphan_span():
+    """Trace chuck does not contain parent span"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "span_not_in_trace", service="service")
+    parent._local_root = parent
+
+    orphan_span = Span(None, "orphan_span", service="service")
+    orphan_span._parent = parent
+    orphan_span._local_root = parent
+    orphan_span.parent_id = orphan_span._parent.span_id
+
+    trace_processors.process_trace([orphan_span])
+    # top_level in orphan_span should be explicitly set to zero/false
+    assert orphan_span.get_metric("_dd.top_level") == 0
+
+
+def test_trace_top_level_span_processor_no_parent_span_object():
+    """Parent span and child span have the different service names and Span._parent is None"""
+    trace_processors = TraceTopLevelSpanProcessor()
+    parent = Span(None, "span_not_in_trace", service="service")
+    parent._local_root = parent
+
+    span_wo_parent_span_obj = Span(None, "span_wo_parent_span_obj", service="new_service_name")
+    span_wo_parent_span_obj.parent_id = parent.span_id
+    span_wo_parent_span_obj._local_root = parent
+    span_wo_parent_span_obj._parent = None
```

Comment:
I don't think this is something that can actually happen in reality.

That is, if you have a hard reference to the parent span as `_local_root` but not `_parent`.

Reply:
Initially this test was designed to mock the flow of distributed traces and partial flushing. I'll remove this test and add a test case which uses `tracer.trace` and `tracer.configure(partial_flush_enabled=True)`

Reply:
debugging I removed it